In [3]:
!pip install pandas

import pandas as pd

titanic_df = pd.read_csv('titanic.csv')

# Pre-Lecture HW

In [5]:
import pandas as pd
titanic_df = pd.read_csv('titanic.csv')

titanic_df.head()

#this shows missing data per column
missing_data_per_column = titanic_df.isnull().sum()
print('Missing data per column:')
print(missing_data_per_column)

#this sums all the missing data
total_missing_values = titanic_df.isnull().sum().sum()
print('Total missing values:')
print(total_missing_values)

#ChatGPT was wrong; this data set does not contain any missing data

Missing data per column:
Survived                   0
Pclass                     0
Name                       0
Sex                        0
Age                        0
Siblings/Spouses Aboard    0
Parents/Children Aboard    0
Fare                       0
dtype: int64
Total missing values:
0


Here's a summary of our conversation so far:

You initially asked for a public funny dataset, and I suggested the Cards Against Humanity dataset, along with a few other humorous options like a Short Jokes dataset and a Dad Jokes dataset, all available on Kaggle.

You then asked for a public dataset with a URL. I provided the Iris Dataset and the Wine Quality Dataset, both from the UCI Machine Learning Repository, along with direct download links.

Next, you requested a public dataset with missing data. I provided the Titanic Dataset, known for missing values in "Age" and "Cabin", and the Ames Housing Dataset, which has missing values in various columns.

Let me know if you'd like further details!

https://chatgpt.com/share/accd373c-8828-4d7b-9efe-117d1b9ad3ec

//

Here's a summary of our conversation so far:

Opening a CSV file in Jupyter Notebook:

You can use the pandas library to open a CSV file by importing it with import pandas as pd and loading the file using pd.read_csv('your_file.csv').
You can then view the data with df.head() (to see the first few rows) or print(df) for the entire DataFrame.
Counting Missing Data in a CSV file:

After loading the CSV into a DataFrame, you can count missing data using:
df.isnull().sum() to get the count of missing values in each column.
df.isnull().sum().sum() to get the total number of missing values across the entire DataFrame.
(df.isnull().sum() / len(df)) * 100 to calculate the percentage of missing values in each column.
Let me know if you need more details or further clarification!

https://chatgpt.com/share/464fd6d6-d5d6-469c-8a30-096e8f2d6525

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()


row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [6]:
df_rows = df.shape[0]
df_columns = df.shape[1]

print('number of rows:' , df_rows)
print('number of columns:' , df_columns)

number of rows: 391
number of columns: 11


### Observations refer to the rows; each row represents a different villager in this dataset. Variables refer to the characteristics of an observation, that is, the columns. So each observation has multiple variables; characteristics of each specific villager in this particular dataset.

In [10]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

df_rows = df.shape[0]
df_columns = df.shape[1]

print('number of rows:' , df_rows)
print('number of columns:' , df_columns)

print(df.info())
print(df.describe())
df.describe(include='all')

number of rows: 391
number of columns: 11
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB
None
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000


row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
max     483.000000      NaN      NaN    NaN     NaN      NaN         NaN   

                song   phrase           full_id  \
count            380      391               391   
unique            92      388               391   
top     K.K. Country  wee one  villager-admiral   
freq              10        2                 1   
mean             NaN      NaN               NaN   
std              NaN      NaN               NaN   
min              NaN      NaN               NaN   
25%              NaN      NaN               NaN   
50%              NaN      NaN               NaN   
75%              NaN      NaN               NaN   
max              NaN      NaN               NaN   

                                                      url  
count                                                 391  
unique                                                391  
top     https://villagerdb.com/images/villagers/thumb/...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [17]:
print(df['species'].value_counts())

print(df.count(axis=0)) #columns
print(df.count(axis=1)) #rows

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64
row_n          391
id             390
name           391
gender         391
species        391
birthday       391
personality    391
song           380
phrase         391
full_id        391
url            391
dtype: int64
0      11
1      11
2      11
3      11
4      11
       ..
386    11
387    11
388    11
389    11
390    11
Length: 391, dtype: int64


### I will now use another dataset with missing numerical and non-numerical data.

In [37]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df_rows = df.shape[0]
df_columns = df.shape[1]
print('number of rows:' , df_rows)
print('number of columns:' , df_columns)

print(df.describe())
print(df.columns)
print(df.columns[3])
print('-------------------------------')
print(df.describe(include = 'all'))

number of rows: 891
number of columns: 15
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')
age
-------------------------------
          survived      pclass 

### Shape and describe methods, attributes and methods

The difference between the number provided by the shape method and the columns presented by the describe method occurs due to the nature of the describe method. It only displays numerical columns by default. In order to display all the columns, including the ones containing non-numerical values, I used the parameter 'include' in order to display all the non-numerical columns. This at the behest of both ChatGPT and Copilot (which I used because I got limited by ChatGPT).

A method is a function inherent to an object, meaning that it can only be called once you have a preexisting object you can call it with. An attribute is data tied to an object, which is why when calling an attribute you do not use parentheses.

### Summary

In this conversation, we discussed the difference between a **method** and an **attribute** in Python:

- **Method**: A function associated with an object, called with parentheses `()` because it can perform actions and accept parameters (e.g., `my_car.start()`).
- **Attribute**: A variable associated with an object, accessed without parentheses, representing properties or data (e.g., `my_car.make`).

Methods define behavior, while attributes represent the object's characteristics.

https://chatgpt.com/share/66e3bd35-cbdc-800b-b46b-be1dfbf029ef

*I forgot to add the definition of shape and attribute, so I had to initiate another conversation with chatgpt for that singular purpose. Above is the summary of that interaction (which was short-lived anyway).*

## .describe() summary statistics

1. Count: Gives the number of values (non-missing values) in a column. If this number is smaller than the total row number, it indicates the existence of missing data.
2. Mean: The average of the values in the column.
3. Std (Standard Deviation): This value represents the tendency of dispersion of values from the mean. This number provides insight into the consistency of the data.
4. Min (Minimum): Lowest value in a given column.
5. Percentiles (25th, 50th, 75th): Provides the boundaries for Q1, the median and the third quartile (Q3). These numbers indicate the spread of the data and help identify outliers. 25th percentile is the number under which 25% of the data falls. The 50th percentile is the number under/over which 50% of the data falls (the midpoint value). The 75th percentile is the number under which 75% of the data falls.
6. Max (Maximum): Highest value in a given column.

In [28]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df_rows = df.shape[0]
df_columns = df.shape[1]
print('number of rows:' , df_rows)
print('number of columns:' , df_columns)
print()
print()



number of rows: 891
number of columns: 15




# Summary of Conversation

## 1. Dataset Contexts
- We initially discussed topics in the context of the **Animal Crossing villagers dataset**.
- Later, we switched to the **Titanic dataset**, which is available at: [Titanic Dataset](https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv).

## 2. Data Exploration
- We explored how to determine the size of a dataset using the `.shape` attribute, which returns a tuple of `(rows, columns)`.
- We discussed the `.describe()` method in pandas, which provides summary statistics of the dataset:
  - **Count**: The number of non-null values.
  - **Mean**: The average value of the column.
  - **Std**: The standard deviation (how spread out the data is).
  - **Min/Max**: The minimum and maximum values.
  - **Percentiles (25%, 50%, 75%)**: Quartiles that split the data into four parts.

## 3. Methods
- We covered the `.dropna()` method in pandas, which removes rows or columns that contain missing (`NaN`) values.
  - You can customize its behavior to drop rows or columns, set thresholds (`thresh`), or specify subsets of data.
  
- We also compared `.dropna()` with Python's `del`:
  - **`.dropna()`**: Removes rows or columns with missing values (pandas-specific).
  - **`del`**: A general Python keyword used to delete variables, list elements, or dictionary keys.

## 4. Printing with Newlines
- We talked about adding space between lines when using the `print()` function by:
  - Using newline characters (`\n`) to add space between lines.
  - Inserting blank `print()` statements to create empty lines between outputs.

---
https://chatgpt.com/share/66e37052-1704-800b-ba42-b069f780992f


# Using Copilot
-------------------------

## Conversation Summary

## Introduction
- **User**: Hi! from now on we will be discussing things in the context of this dataset: Titanic Dataset
- **Assistant**: Acknowledged the dataset and asked what the user would like to explore or analyze first.

## Discussion on `.describe()` Method
- **User**: How does the `.describe()` method work? Why does it provide summaries of numerical values only?
- **Assistant**: Explained that the `.describe()` method provides summary statistics for numerical columns by default, including count, mean, standard deviation, min, 25th percentile, median, 75th percentile, and max. Mentioned that these statistics are most meaningful for numerical data and that the method focuses on numerical values for efficiency and relevance.

## Including All Columns in `.describe()`
- **User**: What does the parameter `include='all'` do in this context?
- **Assistant**: Explained that `include='all'` generates summary statistics for all columns, including categorical ones. For numerical columns, it provides the usual statistics. For categorical columns, it provides count, unique values, top (most frequent) value, and frequency of the top value. Provided an example using the Titanic dataset.

## Request for Summary
- **User**: Can you give me a summary of our conversation so far in Markdown for my Jupyter notebook?
- **Assistant**: Provided the current summary in Markdown format.



# .dropna() vs del

**`.dropna()`:** This method produces a new dataframe without columns or rows that have missing data.

**`del` function:** This function permanently deletes a column from a dataframe.

`.dropna()` allows us to experiment with our dataframe without the risk of permanently altering our data. It also provides the flexibility of eiher getting rid of columns *and* rows.

`del`, on the other and, deletes a column permanently. It is an efficient way of getting rid of data that we know we will not need. It is also useful to free up memory.

So, using `.dropna()` is better when keeping the data in its original state is important, and when we want to manipulate both columns and rows, while `del` is better-suited to handle cases where we need to free up space and we're certain that some data is not required. `del` is also useful when we just want to focus on some columns of the dataset.

# Using `del` and `.dropna()` together

Using `del` before using the `.dropna()` could be useful when dealing with columns you don't need. For example, the `'adult_male'` column could be considered irrelevant when we have another column making the distinction between man and woman passengers. This way, using `.dropna()` could take care of rows with missing data while `del` got rid of a redundant column. 

Another important thing to consider is that calling `.dropna()` to clean rows means that no column would disappear after; no columns would have missing data. Using `del` before would guarantee getting rid of a specific column beforehand.

In [18]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df_rows = df.shape[0]
df_columns = df.shape[1]
print('number of rows:' , df_rows)
print('number of columns:' , df_columns)
print(df.columns)
print()
print()

df_copy = df.copy()
del df['adult_male']
print(df.columns)
print()

df_clean = df.dropna()
print(df_clean)

print()
print()


number of rows: 891
number of columns: 15
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'deck', 'embark_town', 'alive', 'alone'],
      dtype='object')

     survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1           1       1  female  38.0      1      0  71.2833        C  First   
3           1       1  female  35.0      1      0  53.1000        S  First   
6           0       1    male  54.0      0      0  51.8625        S  First   
10          1       3  female   4.0      1      1  16.7000        S  Third   
11          1       1  female  58.0      0      0  26.5500        S  First   
..        ...     ...     ...   ...    ...    ...      ...      ...    ...   
871         1       1  female  47.0      1      1  5

# My approach to the dataset

I used `del` to get rid of a column I deemed irrelevant before cleaning the dataframe using `.dropna()`. While talking to Chatgpt, I realized that if I cleaned the dataframe using `.dropna()` without indicating a specific axis, I would not be able to clean any other columns, as all the missing data is already purged (it might seem stupid, but I had not realized this). In using this method, I keep all the columns except the column I deem useless.

Before, my dataset had lots of missing data and a useless column. After utilizing my (genius) and (clearly) calculated method, my dataset is now clean and without redundant columns.

In [27]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df_rows = df.shape[0]
df_columns = df.shape[1]
print('number of rows:' , df_rows)
print('number of columns:' , df_columns)
print(df.columns)
print()
print('------------------------------------')

df_grouptest = df.groupby('sex')['fare'].describe()
print(df['sex'])
print(df['fare'])
print()
print()
print(df_grouptest)

number of rows: 891
number of columns: 15
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

------------------------------------
0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: sex, Length: 891, dtype: object
0       7.2500
1      71.2833
2       7.9250
3      53.1000
4       8.0500
        ...   
886    13.0000
887    30.0000
888    23.4500
889    30.0000
890     7.7500
Name: fare, Length: 891, dtype: float64


        count       mean        std   min        25%   50%    75%       max
sex                                                                        
female  314.0  44.479818  57.997698  6.75  12.071875  23.0  55.00  512.3292
male    577.0  25.523893  43.138263  0.00   7.895800  10.5  26.55  512.3292


### `.groupby('col1')['col2'].describe()`

What this function does essentially is group the values in column 2 according to column 1. That is, for all the same values in column 1, the function shows the values in column 2 that correspond to them. Chaining `.describe()` gives the statistical summary of the column 2 values corresponding to the groups created from the data in column 1.

In my example above, I used the function to show me the statistical summary of the fares female and male passengers paid in the titanic. In this example, column 1 is 'sex' and column 2 is 'fare'. This shows me the summary of the fare vaules according to whether a passenger was female or male.

### `count` in `.groupby()`

The difference between `.count()` and `count` (the latter being in the `.grouby()` function), is that the former focuses on the data as a whole, while the latter focuses on the data in 'col2' relative to 'col1'.

## Data Analysis Summary

### 1. Difference Between `.dropna()` and `del`

- **`.dropna()`**:
  - Used in pandas to handle missing data by removing rows or columns that contain `NaN` values.
  - Returns a new DataFrame unless `inplace=True` is specified.
  - Customizable using parameters like `axis=0` (for rows) or `axis=1` (for columns).

- **`del`**:
  - A Python function used to delete entire columns or variables.
  - Alters the original DataFrame in place by removing the specified column.
  - Useful for freeing up memory, especially when working with large datasets.

### 2. Dropping Rows and Columns with `.dropna()`

- **Dropping Rows (`axis=0`, default)**:
  Removes rows that contain missing values.
  
  ```python
  df.dropna()  # Drops rows with any NaN values

    
    Dropping Columns (axis=1): Removes columns that contain missing values.

python
Copy code
df.dropna(axis=1)  # Drops columns with any NaN values
Dropping Both Rows and Columns: To drop rows first and then columns:

python
Copy code
df_cleaned = df.dropna().dropna(axis=1)
3. Using .groupby() in pandas
.groupby("col1")["col2"].describe():

Groups the DataFrame by the unique values in col1.
Provides summary statistics for col2 within each group, such as count, mean, std, min, 25%, 50%, 75%, and max.
Example:

python
Copy code
df.groupby("col1")["col2"].describe()
Example output:

arduino
Copy code
       count  mean  std  min  25%  50%  75%  max
col1
A       2.0   1.5   0.7  1.0  1.25 1.5  1.75 2.0
B       2.0   5.5   0.7  5.0  5.25 5.5  5.75 6.0
C       2.0   7.5   0.7  7.0  7.25 7.5  7.75 8.0
4. How .groupby() Assigns Values
Grouping:

.groupby("col1") groups the DataFrame by unique values in col1.
It assigns col2 values to each group based on the corresponding col1 value.
Example:

python
Copy code
grouped = df.groupby("col1")["col2"]

for name, group in grouped:
    print(f"Group: {name}")
    print(group)
Example output:

yaml
Copy code
Group: A
0    1
1    2
Name: col2, dtype: int64

Group: B
2    5
3    6
Name: col2, dtype: int64

Group: C
4    7
5    8
Name: col2, dtype: int64
5. Using .copy()
.copy():

Creates a deep copy of a DataFrame or Series.
Changes made to the copy do not affect the original DataFrame.
python
Copy code
df_copy = df.copy()
This is useful when you want to modify a DataFrame without altering the original data.

csharp
Copy code

You can paste this directly into a markdown cell in Jupyter!

https://chatgpt.com/share/66e3b715-a010-800b-a21a-94972d410ead

*By this point, Chatgpt was already 'hallucinating'.*

# Have I reviewed the course information?

Yes, mostly. I have reviewed material that I couldn't follow during this week's lecture.

Troubleshooting is included in another file, just so I could avoid having to re-run my whole notebook.